sdk/eventgrid/eventgrid/src/cloudEventDistrubtedTracingEnricherPolicy.ts

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import {
+  HttpOperationResponse,
+  RequestPolicy,
+  RequestPolicyOptions,
+  RequestPolicyFactory,
+  WebResourceLike
+} from "@azure/core-http";
+
+export const CloudEventBatchContentType = "application/cloudevents-batch+json; charset=utf-8";
+export const TraceParentHeaderName = "traceparent";
+export const TraceStateHeaderName = "tracestate";
+export const ContentTypeHeaderName = "Content-Type";
+
+export function cloudEventDistributedTracingEnricherPolicy(): RequestPolicyFactory {
+  return {
+    create: (nextPolicy: RequestPolicy, _options: RequestPolicyOptions) => {
+      return new CloudEventDistributedTracingEnricherPolicy(nextPolicy);
+    }
+  };
+}
+
+/**
+ * CloudEventDistributedTracingEnricherPolicy is a policy which adds distributed tracing informatiom
+ * to a batch of cloud events. It does so by copying the `traceparent` and `tracestate` properties
+ * from the HTTP request into the individual events as extension properties.
+ *
+ * This will only happen in the case where an event does not have a `traceparent` defined already. This
+ * allows events to explicitly set a traceparent and tracestate which would be respected during "multi-hop
+ * transmition".
+ *
+ * See https://github.com/cloudevents/spec/blob/master/extensions/distributed-tracing.md
+ * for more information on distributed tracing and cloud events.
+ */
+export class CloudEventDistributedTracingEnricherPolicy implements RequestPolicy {
+  constructor(readonly _nextPolicy: RequestPolicy) {}
+
+  sendRequest(httpRequest: WebResourceLike): Promise<HttpOperationResponse> {
+    if (!httpRequest) {
+      throw new Error("webResource cannot be null or undefined");
+    }
+
+    const traceparent = httpRequest.headers.get(TraceParentHeaderName);
+    const tracestate = httpRequest.headers.get(TraceStateHeaderName);
+
+    if (
+      httpRequest.headers.get(ContentTypeHeaderName) === CloudEventBatchContentType &&
+      typeof httpRequest.body === "string" &&
+      traceparent
+    ) {
+      // per the cloud event batched content type we know the body is an array encoded in JSON.
+      const body = JSON.parse(httpRequest.body) as any[];
+
+      const enrichedBody = body.map((value: any) => {
+        if (typeof value !== "object" || value.traceparent) {
+          return value;
+        }
+
+        const ret = { ...value };
```

Comment:
why bother creating a new object here? `value` was only lately created by `JSON.parse` and nobody else should hold a reference to it. 

I think instead of doing a map here you could simply for/of over the array and conditionally set these properties if needed.

Reply:
> why bother creating a new object here? value was only lately created by JSON.parse and nobody else should hold a reference to it.

Good catch.  This is an artifact of me misunderstanding the `body` parameter (I initially thought it would be a JSON object, not a string) and then when I fixed that I didn't revisit this with a critical eye.

> I think instead of doing a map here you could simply for/of over the array and conditionally set these properties if needed.

SGTM!